Kinetix.Tools.Core + T4 and parameters singleton removal + .NET Core

JabX opened the pull request:
Preparing groundwork for KleeGroup/kinetix-tools#9, KleeGroup/kinetix-tools#10 and KleeGroup/kinetix-tools#11

This is a PR because it's completely untested, though I am confident there won't be any regression 😃 

I created `Kinetix.Tools.Common` where I put all parameter types (`GeneratorParameters` and co.), the `Model` namespace (i needed to use `ModelProperty` so I copied everything over), `FileWriter` in prevision of the TS gen migration, and every TS util we had across both generators.

While working on parameters, I removed the `Singletons.GeneratorParameters` and gave every class the parameters it needed to work instead. I realized we some of them were actually misplaced, so I moved them : 
- **`Kinetix`** from **root** to **csharp**
- **`DefaultValuesFile`**, **`NoTableFile`**, **`HistoriqueCreationFile`** from **root** to **ssdt**

I removed the T4 files and kept the generated C#, that I cleaned up by removing all the useless things that were generated. It is perfectly usable as it is, but it would be better if we could build in the future something like `CSharpWriter` on top of it.

Again, a lot of stuff moved around, because that's my favorite thing to do 😛 

Please update your config file and test the PR to confirm there's no regression on your side (well, you could wait for me to test it first on Monday).

(Also, I might have changed the output folder of the generators to remove the framework prefix, so we can more easily test between full .NET and Core in the future.)

JabX commented:
I checked on my project and it works well.

I added a new feature to use the default values from the OOM to generate the adequate C# for EF and the default value in the generated SQL. I think it should work with SSDT too, with the default value file overriding the default values set in the OOM.

I also removed the MSBuild dependency for SSDT as discussed (since wildcards work!)

Finally, I rebased (painfully) on master, so please check that everything still works fine for you, and you can merge.

JabX commented:
I managed to get both generators work on .NET Core with minimal changes.
You'll need the 2.1 SDK installed (it doesn't come with VS yet) and to change your generators calls from `%modgen%.exe` to `dotnet %modgen%.dll`.

c3dr0x commented:
So, I tried It.

### Class generator

Almost ok, it changed the content of my sqlproj file by what seems to be the content of a list factory sql file

```xml
----
---- ATTENTION CE FICHIER EST GENERE AUTOMATIQUEMENT (7667C01C58993CD698162FA04A2840CE180E56F7) !
----

-- ===========================================================================================
--   Description		:	Insertion des valeurs de listes statiques.
-- ===========================================================================================


```

### Service generator

It doesn't run:

![image](https://user-images.githubusercontent.com/1591017/41910871-40802ebc-794b-11e8-9943-c66ad427dba8.png)

JabX commented:
The ModGen issue is weird and I'd appreciate if you could take a look at what's happening (since you have a SSDT project on hand). It should be pretty trivial to find the problem and fix it.

And I'm pretty sure the ServiceGen issue comes from the fact that your project is a classic ASP.NET project that the .NET Core SDK cannot open... Which is probably (one of the reasons) why the classic `MSBuildWorkspace` isn't supported on Core. I'm gonna look into it when I have the time to try and find a solution for this (which doesn't force us to go back to full .NET)

c3dr0x commented:
Okay I found why it changed the .sqlproj file

https://github.com/get-focus/kinetix-tools/blob/preemptive-refactor/Kinetix.ClassGenerator/SsdtSchemaGenerator/SqlScriptEngine.cs#L100

JabX commented:
The sqlproj problem should be fixed now, and I added the option to multitarget net471 and netcoreapp2.1 because using `MSBuildWorkspace` is mandatory to handle non-Core projects and isn't available on Core.

I also noticed that Buildalyzer wasn't properly emulating MSBuildWorkspace on Core because it didn't fill the `Folders` property on each `Document`. So I added a workaround for it (specific to Core) and fixed the code so that the behavior (and the output) is identical between both versions. Meaning I can use the net471 version to generate for a Core project (netcoreapp2.1 will be mandatory when we'll want to package it as a tool but we're not here yet). 

Please let me know if it works fine for you, so that we can merge it and go forward with new changes.

JabX commented on Kinetix.SpaServiceGenerator/Program.cs:
```diff
@@ -61,9 +74,18 @@ internal class Program
-                var secondFolder = controller.Folders.Count > 1 ? $"/{string.Join("/", controller.Folders.Skip(1).Select(f => f.ToDashCase()))}" : string.Empty;
-                var folderCount = (frontEnds.Count() > 1 ? 1 : 0) + controller.Folders.Count;
+                var secondFolder = folders.Count > 1 ? $"/{string.Join("/", folders.Skip(1).Select(f => f.ToDashCase()))}" : string.Empty;
+                var folderCount = (frontEnds.Count() > 1 ? 1 : 0) + folders.Count - 1;
```
This is used to generate the JS imports in the service files, and I noticed a "-1" was missing (my `Folders.Count` was always 0 instead of always 1, so I didn't catch it).
I hope this works for your folder hierarchy
